### DISCUSSION TITLE
ListItems going out of view

### DISCUSSION THREAD

**Discussion by AnthonyTschirhard** [Textualize/textual#4851] [Q&A] [answered]
Hello everyone,

I'm posting this question here because I've read the documentation multiple times (EXCELLENT doc and code docstrings by the way, it's a pleasure to read the friendly manual) but I'm having a problem I can't get my head around.

I have a custom TaskList widget class that inherits from ListView. And when I scroll down (with "j"), at some point the ListItem object goes out of view.

```python
from textual.app import App, ComposeResult
from textual.screen import Screen
from textual.widgets import Header, Footer, Static, ListView, ListItem

class TaskList(ListView):
    def __init__(self):
        self.items = [ListItem(Static(f"task {i}")) for i in range(100)]
        super().__init__(*self.items)
        self.index = 0

    def next_task(self):
        self.index += 1

class MyScreen(Screen):
    widgets: list

    def __init__(self):
        self.widgets = []
        self.index_widgets = -1

        super().__init__()

    def compose(self) -> ComposeResult:
        sprint = TaskList()
        self.widgets.append(sprint)
        yield sprint

    def next_task(self):
        self.widgets[self.index_widgets].next_task()


class LazyJiraApp(App):
    CSS = """
        MyScreen {
            layout: grid;
        }
    """

    BINDINGS = [
        ("q", "quit", "Quit the app"),
        ("j", "next_task", "Next Task"),
    ]

    def __init__(self):
        super().__init__()
        self.my_screen = MyScreen()

    def compose(self) -> ComposeResult:
        yield self.my_screen

    def action_next_task(self) -> None:
        self.my_screen.next_task()
        self.refresh()

if __name__ == "__main__":
    app = LazyJiraApp()
    app.run()
```

Would you have an idea what I'm missing? If you need more context about the code (work in progress, you can't run it presently): [LazyJira](https://github.com/AnthonyTschirhard/lazyjira).
Thanks a lot!
Anthony

**Comment by TomJGooding** [comment]
The docs for the `refresh` method includes this warning:

> It is rarely necessary to call this method explicitly. Updating styles or reactive attributes will do this automatically.

I'm not sure exactly what you are trying to achieve here with your custom ListView?

**Reply by AnthonyTschirhard** [reply]
It's rarely useful to use the refresh method when you modify a reactive attribute, which is the case here since I'm modifying the index (listview.index is reactive). But my problem is that the self.scroll_to_center doesn't refresh my view when it's called (and the self.refresh doesn't have an impact).
To answer your question, I'm using a customer ListView because I need to extend its functionalities (but I guess my problem would be the same with a ListView)

**Reply by TomJGooding** [reply]
Could you clarify exactly what you expect your `previous/next_task` methods to do? For example here's my quick attempt at an MRE which hopefully will help illustrate:

```python
from textual.app import App, ComposeResult
from textual.widgets import Footer, Label, ListItem, ListView


class TaskListView(ListView):
    def __init__(self) -> None:
        self.tasks = [ListItem(Label(f"Task #{i}")) for i in range(1, 101)]
        super().__init__(*self.tasks)

    def next_task(self) -> None:
        if self.index is not None:
            self.index += 1

    def previous_task(self) -> None:
        if self.index is not None:
            self.index -= 1


class ExampleApp(App):
    BINDINGS = [
        ("j", "next_task", "next_task"),
        ("k", "previous_task", "previous_task"),
    ]

    CSS = """
    TaskListView {
        height: 10;
    }
    """

    def compose(self) -> ComposeResult:
        yield TaskListView()
        yield Footer()

    def action_next_task(self) -> None:
        task_list = self.query_one(TaskListView)
        task_list.next_task()

    def action_previous_task(self) -> None:
        task_list = self.query_one(TaskListView)
        task_list.previous_task()


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Reply by AnthonyTschirhard** [reply]
Thank you so much for taking the time, I really appreciate it!
And I've also realized that I was absolutely not clear in my initial post. By the way, your code behaves exactly like I hope mine would. The selected items in your list never go "out of view". Mine do. And because your TaskList Object is not so different from mine, I think the problem comes from somewhere else (maybe a "scroll" css option I missed?).

This is roughly my code: 
[main.txt](https://github.com/user-attachments/files/16553936/main.txt)

(sorry I couldn't upload the .py)

As you can see, when you go down with "j", the selected item goes out of view. That's why I thought I needed some kind of refresh ; if you focus out of the terminal, the selected item will "come back in view".

**Reply by TomJGooding** [reply]
Rather than just posting your code, I'd recommend trying to create a [minimal reproducible example](https://github.com/Textualize/textual/discussions/3361). Perhaps start with my example code which you found works, then add your functionality until it breaks.

**Reply by AnthonyTschirhard** [reply]
I didn't post my code, I took some time to refactor it so that it is a minimal reproducible example :-)
It may seem like a long code because I've put the css that I use inside, so that you don't have to download 2 files. I could remove the function related to the h,j,k,l moves because they take some space, but then you could not observe the problem.

**Reply by AnthonyTschirhard** [reply]
There is one line to remove though, sorry about that:
from task import BaseTask
(you don't need it in this example)

**Reply by TomJGooding** [reply]
There's a number of issues with your example - incorrect CSS, [managing screens](https://textual.textualize.io/guide/screens/), etc - amd also a lot of unnecessary code that doesn't help demonstrate your issue. Please could you try fixing these issues first, so people trying to help can at least run your app?

**Reply by TomJGooding** [reply]
If all you are trying to do is add new bindings to focus the next/previous widget or list item, there's a much easier way by utilising the existing `action` methods than what it looks like you're doing here. For example:

```python
from textual.app import App, ComposeResult
from textual.containers import Horizontal
from textual.screen import Screen
from textual.widgets import Footer, Label, ListItem, ListView


class TaskListView(ListView):
    BINDINGS = [
        ("k", "cursor_up", "Previous Task"),
        ("j", "cursor_down", "Next Task"),
    ]

    DEFAULT_CSS = """
    TaskListView {
        border: tall $panel-lighten-3;

        &:focus {
            border: tall $accent;
        }
    }
    """

    def __init__(self) -> None:
        self.tasks = [ListItem(Label(f"Task #{i}")) for i in range(1, 51)]
        super().__init__(*self.tasks)


class KanbanScreen(Screen):
    BINDINGS = [
        ("h", "app.focus_previous", "Previous Group"),
        ("l", "app.focus_next", "Next Group"),
    ]

    def compose(self) -> ComposeResult:
        to_do = TaskListView()
        to_do.border_title = "TODO"

        doing = TaskListView()
        doing.border_title = "DOING"

        done = TaskListView()
        done.border_title = "DONE"

        with Horizontal():
            yield to_do
            yield doing
            yield done

        yield Footer()


class KanbanApp(App):
    def on_mount(self) -> None:
        self.push_screen(KanbanScreen())


if __name__ == "__main__":
    app = KanbanApp()
    app.run()
```

**Reply by AnthonyTschirhard** [reply]
Hi,

You are perfectly right, I could have explained my problem with a code way shorter. Here is a version stripped out of everything unnecessary (you can only go down with the tasks though):
[main.txt](https://github.com/user-attachments/files/16570125/main.txt) => I think you'll like this one ;-)
You will see that when you go down the tasks, eventually the selected task will disappear.

Also, I absolutely love the binding:
        ("h", "app.focus_previous", "Previous Group"),
The reason I didn't choose this approach is that I only want to circle between certain widgets only. But I've just realized that I could still use your example and make some widgets non-focusable.

**Reply by AnthonyTschirhard** [reply]
Also, would you mind telling me why my css was wrong? I've read the documentation you've sent me quite a few times but I must have missed something.

I've also updated the title and content of the discussion accordingly.

**Reply by TomJGooding** [reply]
> Also, I absolutely love the binding:
("h", "app.focus_previous", "Previous Group"),
The reason I didn't choose this approach is that I only want to circle between certain widgets only. 

Thanks for clarifying, perhaps instead you could use [`Screen.focus_next`](https://textual.textualize.io/api/screen/#textual.screen.Screen.focus_next) which you can filter by a selector?

```python
class KanbanScreen(Screen):
    BINDINGS = [
        ("l", "focus_next_group", "Next Group"),
    ]

    def action_focus_next_group(self) -> None:
        self.focus_next(TaskListView)
```

**Reply by AnthonyTschirhard** [reply]
Oh yes that's absolutely perfect (and elegant).

Still, now I see this light version of the code, I understand even less why my ListItems are going out of view... Even if I completely remove the css, I still have the problem. I don't seen between your first code example and this, why your listitems stay in view and mine don't.

**Reply by TomJGooding** [reply, marked as answer]
I've only skimmed your trimmed down example, but you shouldn't be yielding screens. making lists of widgets, or explicitly calling refresh.
Hopefully I've shown that Textual has better ways of handling all of this?

**Reply by AnthonyTschirhard** [reply]
Hey,
Thanks a lot, you've definitely shown me a better way of using textual ;-) My code is so much cleaner now.
As soon as I stopped yielding the screen, everything was fixed. I wonder why an exception is not raised when you try to yield a screen but I guess there's probably a very good reason for that!
Thanks again!!
